### PR TITLE
Implement splitting of ratings

### LINKED ISSUE #108: Implement splitting of ratings
See `create_split()` in `items/ratings.py` and `main()` in `run_simulation.py`.

Add `historical_ratings_ratio` to config with a default value of 0.8.

### PR DESCRIPTION
**Pull request by NoB0**
Implement `create_split` method and add tests.

Fixes #108

### CODE REVIEW

File: usersimcrs/items/ratings.py [hunk 1]
```diff
@@ -125,13 +141,31 @@ def create_split(
...
+
+        # Determine the number of items to use as historical data.
+        nb_historical_items = int(historical_ratio * len(self._item_ratings))
+
         historical_ratings = Ratings(self._item_collection)
         ground_truth_ratings = Ratings(self._item_collection)
-        # TODO: Implement this method with tests
-        # See: https://github.com/iai-group/UserSimCRS/issues/108
+
+        for i, (item_id, user_ratings) in enumerate(self._item_ratings.items()):
```
Comment (kbalog) [inline usersimcrs/items/ratings.py:160]: Here, we need to _randomly_ sample x% of the ratings _per user_ as historical data. For testability, it'd be good for this method to take a user_item_sampler Callable argument. (That way, we could have another sampler later if we want the first x% ratings based on timestamp as historical data, as opposed to random x%.)

**Review by kbalog — CHANGES_REQUESTED** [review]
Sampling of historical ratings needs to be done slightly differently, see comment.

File: usersimcrs/items/ratings.py (hunk 1 above)
Comment (NoB0) [reply usersimcrs/items/ratings.py]: I am not entirely sure how you see the implementation of this sampler. I made a suggestion that is open for open discussion.

File: usersimcrs/items/ratings.py [hunk 2]
```diff
@@ -133,13 +166,16 @@ def get_random_user_id(self) -> str:
...
+        self,
+        historical_ratio: float,
+        sampler: Callable = user_item_sampler,
     ) -> Tuple["Ratings", "Ratings"]:
         """Splits ratings into historical and ground truth ratings.
 
         Args:
             historical_ratio: Ratio ([0..1]) of ratings to be used as historical
               data.
+            sampler: Callable performing the sampling.
```
Comment (kbalog) [inline usersimcrs/items/ratings.py:178]: Nit: "... the sampling of items per user."

File: usersimcrs/items/ratings.py [hunk 3]
```diff
@@ -151,21 +187,11 @@ def create_split(
...
-                    historical_ratings.add_user_item_rating(
-                        user_id, item_id, rating
-                    )
-                else:
-                    ground_truth_ratings.add_user_item_rating(
-                        user_id, item_id, rating
-                    )
+        sampler(
+            self, historical_ratio, historical_ratings, ground_truth_ratings
+        )
```
Comment (kbalog) [inline usersimcrs/items/ratings.py:195]: The outer loop over the users should be here, and the sampler samples ratings for a given user.

File: usersimcrs/items/ratings.py [hunk 4]
```diff
@@ -11,13 +11,46 @@
...
 logger = logging.getLogger(__name__)
 
 
+def user_item_sampler(
+    ratings: "Ratings",
+    historical_ratio: float,
+    historical_ratings: "Ratings",
+    ground_truth_ratings: "Ratings",
+) -> None:
+    """Creates a random sample of items for each users.
```
Comment (kbalog) [inline usersimcrs/items/ratings.py:27]: This should create the sample of items for a given user.

**Review by kbalog — CHANGES_REQUESTED** [review]
`user_item_sampler()` should sample items only for a given user, not all users. There should also be tests for it, which is easier if sampling is done for a single user.

When testing the `create_split()` method, it could take a simpler sampler that deterministically splits items (e.g., first x% of the list, no randomization), so that we could also test whether individual items end up in the correct split.

File: usersimcrs/items/ratings.py (hunk 2 above)
Comment (NoB0) [reply usersimcrs/items/ratings.py]: Done.

**Comment by NoB0** [comment]
I refactored `user_item_sampler()` and updated `create_split()` accordingly.

File: tests/items/test_ratings.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+}
+ITEMS_CSV_FILE = "tests/data/items/movies.csv"
+RATINGS_CSV_FILE = "tests/data/items/ratings.csv"
+
+
+def simple_user_item_sampler(
+    item_ratings: Dict[str, float],
+    historical_ratio: float,
+) -> List[str]:
+    """Creates a sample of items for a given user.
```
Comment (kbalog) [inline tests/items/test_ratings.py:28]: Nit: Can you mention that it takes the first portion of ratings as opposed to a random sample?

File: tests/items/test_ratings.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    ratings = Ratings()
+    ratings.load_ratings_csv(RATINGS_CSV_FILE)
+    return ratings
+
+
+@pytest.mark.parametrize(
+    "historical_ratio",
+    [0.5, 0.8, 0.2],
+)
+def test_user_item_sampler(historical_ratio: float, ratings: Ratings) -> None:
```
Comment (kbalog) [inline tests/items/test_ratings.py:63]: Nit: add `_ratio` to the method name?

File: tests/items/test_ratings.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        is None
+    )
+
+    assert ground_truth_ratings.get_user_item_rating(
+        user_id, ground_truth_item_id
+    )
+    assert (
+        historical_ratings.get_user_item_rating(user_id, ground_truth_item_id)
+        is None
+    )
```
Comment (kbalog) [inline tests/items/test_ratings.py:119]: The second asserts might be removed (disjoint test ensures that those would hold)

**Review by kbalog — APPROVED** [review]
LGTM

File: tests/items/test_ratings.py (hunk 6 above)
Comment (NoB0) [reply tests/items/test_ratings.py]: Done.

File: tests/items/test_ratings.py (hunk 7 above)
Comment (NoB0) [reply tests/items/test_ratings.py]: Asserts removed.

**Pull request merged**